src/graph/network/socket_communicator.cc

```diff
@@ -0,0 +1,191 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file communicator.cc
+ * \brief SocketCommunicator for DGL distributed training.
+ */
+#include <dmlc/logging.h>
+
+#include "socket_communicator.h"
+#include "../../c_api_common.h"
+
+namespace dgl {
+namespace network {
+
+const int kTimeOut = 5;  // 5 minutes for socket timeout
+const int kMaxConnection = 1024;  // 1024 maximal socket connection
+const int kMaxBuffer = 200 * 1024 * 1024;  // 200 MB bufferfor each message
+
+bool SocketCommunicator::Initialize(bool is_sender,
+                                    const char* ip,
+                                    int port,
+                                    int num_sender,
+                                    int queue_size) {
+  if (is_sender) {
+    is_sender_ = true;
+    return InitSender(ip, port);
+  } else {
+    is_sender_ = false;
+    return InitReceiver(ip, port, num_sender, queue_size);
+  }
+}
+
+bool SocketCommunicator::InitSender(const char* ip, int port) {
+  // Sender only has a client socket
+  socket_.resize(1);
+  socket_[0] = new TCPSocket();
```

Comment:
why not use shared_ptr?

Reply:
I think it is better to control the pointer by handle in Finalize() method.